apm-agent-core/src/test/java/co/elastic/apm/agent/bci/InstrumentationTest.java

```diff
@@ -209,6 +276,135 @@ String exceptionPlease() {
+            ByteBuddyAgent.install(),
+            Collections.singletonList(new CommonsLangInstrumentation()));
+
+        assertThat(CommonsLangInstrumentation.enterCount).hasValue(0);
+        assertThat(CommonsLangInstrumentation.exitCount).hasValue(0);
+
+        assertThat(StringUtils.startsWithIgnoreCase("APM", "apm")).isTrue();
+
+        assertThat(CommonsLangInstrumentation.enterCount).hasPositiveValue();
+        assertThat(CommonsLangInstrumentation.exitCount).hasPositiveValue();
+    }
+
+    @Test
+    void testPatchClassFileVersionJava5ToJava7() {
+        // loading classes compiled with bytecode level 49 (Java 6)
+        new org.slf4j.event.SubstituteLoggingEvent();
+
+        // retransforming classes and patch to bytecode level 51 (Java 7)
+        ElasticApmAgent.initInstrumentation(tracer,
+            ByteBuddyAgent.install(),
+            Collections.singletonList(new LoggerFactoryInstrumentation()));
+
+        assertThat(LoggerFactoryInstrumentation.enterCount).hasValue(0);
+        assertThat(LoggerFactoryInstrumentation.exitCount).hasValue(0);
+
+        new org.slf4j.event.SubstituteLoggingEvent();
+
+        assertThat(LoggerFactoryInstrumentation.enterCount).hasPositiveValue();
+        assertThat(LoggerFactoryInstrumentation.exitCount).hasPositiveValue();
+    }
+
+    @Test
+    void testPatchClassFileVersionJava5ToJava7CommonsMath() {
+        org.apache.commons.math3.stat.StatUtils.max(new double[]{3.14});
+
+        // retransforming classes and patch to bytecode level 51 (Java 7)
+        ElasticApmAgent.initInstrumentation(tracer,
+            ByteBuddyAgent.install(),
+            Collections.singletonList(new StatUtilsInstrumentation()));
+
+        assertThat(StatUtilsInstrumentation.enterCount).hasValue(0);
+        assertThat(StatUtilsInstrumentation.exitCount).hasValue(0);
+
+        org.apache.commons.math3.stat.StatUtils.max(new double[]{3.14});
+
+        assertThat(StatUtilsInstrumentation.enterCount).hasPositiveValue();
+        assertThat(StatUtilsInstrumentation.exitCount).hasPositiveValue();
+    }
+
+    @Test
+    void testPrivateConstructorJava7() {
+        org.apache.commons.pool2.impl.CallStackUtils.newCallStack("", false, false);
+
+        // retransforming classes and patch to bytecode level 51 (Java 7)
+        ElasticApmAgent.initInstrumentation(tracer,
+            ByteBuddyAgent.install(),
+            Collections.singletonList(new CallStackUtilsInstrumentation()));
+
+        assertThat(CallStackUtilsInstrumentation.enterCount).hasValue(0);
+        assertThat(CallStackUtilsInstrumentation.exitCount).hasValue(0);
+
+        org.apache.commons.pool2.impl.CallStackUtils.newCallStack("", false, false);
+
+        assertThat(CallStackUtilsInstrumentation.enterCount).hasPositiveValue();
+        assertThat(CallStackUtilsInstrumentation.exitCount).hasPositiveValue();
+    }
+
+    @Test
+    void testPluginClassLoaderGCdAfterUndoingInstrumentation() {
```

Comment:
😍 
Awesome test. 
It verifies that reverting the instrumentation allows the plugin class loader to be GCed.

We also need a test that having the instrumentation in place doesn't prevent from the target class loader and the classes it loaded from being GCed and that when they are GCed, the plugin class loader and plugin classes are GCed as well. A relevant scenario is undeploying an application from a servlet container.
For example: https://github.com/elastic/apm-agent-java/blob/842185664456651f1d6a99602d21a7c052ddf111/apm-agent-core/src/test/java/co/elastic/apm/agent/bci/HelperClassManagerTest.java#L147

Reply:
When this one fails, do we have an easy way to track what is not being GCed ? If a memory dump is required why not capture one and invite user to analyze it ? (that might not be a very good idea on CI).

Reply:
The way to debug it would be to reproduce locally, set a breakpoint after `System.gc()` and to take a heap dump.

Reply:
The new test seems to cover what we need 👍